src/sliceList.ml

```diff
@@ -0,0 +1,195 @@
+    Some (element, list)
+  | None -> (None : (slice_list_element * slice_list) option)
+
+(* Updates the burrow slices in the provided auction state using the given burrow slice list *)
+let slice_list_to_auction_state (auctions: liquidation_auctions) (l: slice_list) =
+  match l with SliceList meta ->
+    let burrow_liquidation_slice = match meta.slice_list_bounds with
+      | None -> (None: burrow_liquidation_slices option)
+      | Some bounds -> (Some {
+          youngest_slice=bounds.slice_list_youngest_ptr;
+          oldest_slice=bounds.slice_list_oldest_ptr;
+        })
+    in
+    let burrow_slices = Ligo.Big_map.update meta.slice_list_burrow burrow_liquidation_slice auctions.burrow_slices in
+    {auctions with burrow_slices = burrow_slices;}
+
+
+(* Appends a new element to the list. This element will be the "youngest" one in the list.
+   You must specify an avl root which this new element will reside under along with the
+   end of the avl queue which you would like to place the element at.
+*)
+let slice_list_append (l:slice_list) (storage:mem) (root:liquidation_auction_id) (queue_end:queue_end) (slice_contents:liquidation_slice_contents) =
+  let meta = match l with SliceList m -> m in
+  (* FIXME: Perhaps throw specific error code here? *)
+  assert (slice_contents.burrow = meta.slice_list_burrow);
+  match meta.slice_list_bounds with
+  (* List is empty, creating the first element *)
+  | None ->
+    let slice = {younger=(None: leaf_ptr option); older=(None: leaf_ptr option); contents=slice_contents;} in
+    (* Write slice to AVL backend *)
+    let storage, ptr = match queue_end with
+      | Back -> avl_push_back storage root slice
+      | Front -> avl_push_front storage root slice
+    in
+    let bounds = {
+      slice_list_youngest_ptr=ptr;
+      slice_list_oldest_ptr=ptr;
+    }
+    in storage, SliceList {meta with slice_list_bounds=Some bounds;}, SliceListElement (ptr, slice)
+  (* The list already has some elements. Need to do some updating.*)
+  | Some bounds ->
+    (* The new element is now the youngest *)
+    let slice = {younger=(None: leaf_ptr option); older=Some bounds.slice_list_youngest_ptr; contents=slice_contents;} in
+    (* Write slice to AVL backend *)
+    let storage, ptr = match queue_end with
+      | Back -> avl_push_back storage root slice
+      | Front -> avl_push_front storage root slice
+    in
+    (* Touch up the old element in the backend *)
+    let former_youngest = bounds.slice_list_youngest_ptr in
+    let storage = avl_update_leaf storage former_youngest (fun (s: liquidation_slice) -> {s with younger = Some ptr}) in
+    (* Update up the bounds with the new youngest element *)
+    let bounds = {bounds with slice_list_youngest_ptr = ptr} in
+    storage, SliceList {meta with slice_list_bounds=Some bounds;}, SliceListElement (ptr, slice)
+
+(* Remove the element from the list, returning its contents *)
+let slice_list_remove (l:slice_list) (storage:mem) (e:slice_list_element) =
```

Comment:
I think I would appreciate annotating the return types at top-level functions. My first instinct to understand a function is usually looking at the type signature, but here it is missing half the information.

Interesting thing is that `ligo` used to require having full type signatures on functions, I guess that requirement is relaxed at one point.